src/state.rs

```diff
@@ -91,127 +69,133 @@ impl State {
             },
             genesis: params.genesis,
             status: Status::Instantiated,
-            checkpoints: empty_checkpoint_map,
-            stake: empty_stake_map,
-            window_checks: empty_votes_map,
+            checkpoints: TCid::new_hamt(store)?,
+            stake: TCid::new_hamt(store)?,
+            window_checks: TCid::new_hamt(store)?,
             validator_set: Vec::new(),
-            testing: is_test,
             expected_msg: Vec::new(),
-        }
+        };
+
+        Ok(state)
     }
 
     /// Adds stake from a validator
-    pub(crate) fn add_stake(
+    pub(crate) fn add_stake<BS: Blockstore>(
         &mut self,
+        store: &BS,
         addr: &Address,
         net_addr: &str,
         amount: &TokenAmount,
     ) -> anyhow::Result<()> {
         // update miner stake
-        let mut bt = make_map_with_root::<_, BigIntDe>(&self.stake, &Blockstore)?;
-        let mut stake = get_stake(&bt, addr)
-            .map_err(|e| anyhow!(format!("error getting stake from Hamt: {:?}", e)))?;
-        stake += amount;
-        set_stake(&mut bt, addr, stake.clone())?;
-        self.stake = bt.flush()?;
-
-        // update total collateral
-        self.total_stake += amount;
-
-        // check if the miner has coollateral to become a validator
-        if stake >= self.min_validator_stake
-            && (self.consensus != ConsensusType::Delegated || self.validator_set.len() < 1)
-        {
-            self.validator_set.push(Validator {
-                addr: addr.clone(),
-                net_addr: String::from(net_addr),
-            });
-        }
-
-        Ok(())
-    }
-
-    pub(crate) fn rm_stake(&mut self, addr: &Address, amount: &TokenAmount) -> anyhow::Result<()> {
-        // update miner stake
-        let mut bt = make_map_with_root::<_, BigIntDe>(&self.stake, &Blockstore)?;
-        let stake = get_stake(&bt, addr)
-            .map_err(|e| anyhow!(format!("error getting stake from Hamt: {:?}", e)))?;
-        // funds being returned
-        let mut stake = stake / LEAVING_COEFF;
-        stake -= amount;
-        set_stake(&mut bt, addr, stake.clone())?;
-        self.stake = bt.flush()?;
-
-        // update total collateral
-        self.total_stake -= amount;
+        let mut bt: TCid<THamt<Cid, TokenAmount>> = TCid::new_hamt(store)?;
```

Comment:
You shouldn't need to create here a new HAMT. With `THamt` you can directly use the type stored in state below:
```rust
self.stake.modify(store,  ...
```